### DISCUSSION TITLE
Can we enforce dependency constraints with `rye sync` and similar?

### DISCUSSION THREAD

**Discussion by jeertmans** [astral-sh/rye#1027] [Q&A]
Hello,

I maintain a Python package that has dependency conflicts for Python>=3.12, but those can be solved using custom dependency constraints or enforce custom dependency resolution, which I could do with PDM.

Is it possible to that with Rye? I could not find a similar issue or help in the docs.

**Comment by bluss** [comment]
rye supports custom environment markers like python_version etc, they need to be written into the pyproject.toml directly (using rye add).

For example like this:

```toml
dependencies = [
    "tomli>=2.0.1; python_version < '3.11'",
]
```

Any more details on what exactly you can do with pdm and what it looks like?

**Reply by jeertmans** [reply]
Maybe my question was not clear, so let me rephrase :-)
My package depends on `A` and `B`. Both `A` and `B` depend on package `C`. E.g., `A` needs `C<1` and `B` needs `C>=1`.

There is a conflict, but I know that setting `C==1` works.

With PDM, I could specify this with:

```toml
[tool.pdm.resolution.overrides]
manimpango = "<1.0.0,>=0.5.0"  # This conflicts with ManimGL, hopefully not an issue
skia-pathops = "0.8.0.post1"  # From manim 0.18.0 (Python 3.12 support)
```

`uv` supports using `--constraint` argument when compiling the lockfile, but `rye sync` does not seem to offer this possibility.

**Reply by bluss** [reply]
I thought rye had something like this but the only "special" option in rye add is --excluded (which does not help you).  However, what about this PR? https://github.com/astral-sh/rye/pull/1015

**Reply by jeertmans** [reply]
Yeah that would probably be a way to fix my issue :) !
